File: emulador_pic_k150/emulador_k150.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstring>
#include <algorithm>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/select.h>
#include <cstdint>
#include <climits>

enum State {
    AWAITING_JUMP_TABLE,
    JUMP_TABLE
};

bool read_exactly(int fd, uint8_t* buf, size_t n) {
    size_t total = 0;
    while (total < n) {
        fd_set rfds;
        FD_ZERO(&rfds);
        FD_SET(fd, &rfds);
        
        struct timeval tv;
        tv.tv_sec = 2;
        tv.tv_usec = 0;
        
        int s = select(fd + 1, &rfds, nullptr, nullptr, &tv);
        if (s <= 0) {
            return false; // Timeout or error
        }
        
        ssize_t r = read(fd, buf + total, n - total);
        if (r <= 0) {
            return false; // EOF or error
        }
        total += r;
    }
    return true;
}

void run_emulator(const std::string& symlink_path) {
    // Initialize mock memory by default (PIC16F628A)
    size_t rom_size_words = 2048;
    size_t rom_size_bytes = rom_size_words * 2; // 4096
    size_t eeprom_size_bytes = 128;
    
    std::vector<uint8_t> rom(rom_size_bytes, 0xFF);
    std::vector<uint8_t> eeprom(eeprom_size_bytes, 0xFF);
    std::vector<uint16_t> fuses_vector(1, 0x3FFF);
    std::vector<uint8_t> pic_id(4, 0xFF);
    uint16_t calibrate = 0x3FFF;
    uint16_t chip_id = 4192;  // PIC16F628A (0x1060)
 
    std::cout << "================ K150 PROGRAMMER EMULATOR (C++) ================" << std::endl;
    std::cout << "Simulating PIC microcontrollers (Dynamic Family Support)" << std::endl;
    std::cout << "Press Ctrl+C to exit emulator" << std::endl;
    std::cout << "================================================================" << std::endl;
 
    while (true) {
        // Create a new pty pair
        int master_fd = posix_openpt(O_RDWR | O_NOCTTY);
        if (master_fd < 0) {
            std::cerr << "Error: posix_openpt failed." << std::endl;
            break;
        }
        if (grantpt(master_fd) < 0 || unlockpt(master_fd) < 0) {
            std::cerr << "Error: grantpt or unlockpt failed." << std::endl;
            close(master_fd);
            break;
        }
        
        char* slave_name = ptsname(master_fd);
        if (!slave_name) {
            std::cerr << "Error: ptsname failed." << std::endl;
            close(master_fd);
            break;
        }
        // Configure slave to be RAW (no ECHO, no ICANON, no flow control)
        int slave_fd = open(slave_name, O_RDWR | O_NOCTTY);
        if (slave_fd >= 0) {
            struct termios attrs;
            if (tcgetattr(slave_fd, &attrs) == 0) {
                attrs.c_iflag &= ~(IXON | IXOFF | IXANY);
                attrs.c_lflag &= ~(ECHO | ICANON | IEXTEN | ISIG);
                attrs.c_oflag &= ~OPOST;
                tcsetattr(slave_fd, TCSANOW, &attrs);
            }
        } else {
            std::cerr << "Warning: Could not open slave to set attributes." << std::endl;
        }
        
        // Update symlink
        unlink(symlink_path.c_str());
        if (symlink(slave_name, symlink_path.c_str()) < 0) {
            std::cerr << "Error: Creating symlink failed." << std::endl;
            close(master_fd);
            break;
        }
        
        std::cout << "\n[K150-EMULATOR-C++] Active virtual port: " << slave_name << std::endl;
        std::cout << "[K150-EMULATOR-C++] Symlink updated: " << symlink_path << std::endl;
        std::cout << "[K150-EMULATOR-C++] Ready. Virtual port symlink: " << symlink_path << std::endl;
        std::cout << "[K150-EMULATOR-C++] Waiting for client connection..." << std::endl;
        
        State state = AWAITING_JUMP_TABLE;
        
        try {
            while (true) {
                uint8_t b;
                ssize_t r = read(master_fd, &b, 1);
                if (r <= 0) {
                    std::cout << "[K150-EMULATOR-C++] Client disconnected (EOF)." << std::endl;
                    break;
                }
                
                // Exit command (b'\x01') always resets state and replies 'Q'
                if (b == 1) {
                    uint8_t resp = 'Q';
                    write(master_fd, &resp, 1);
                    state = AWAITING_JUMP_TABLE;
                    continue;
                }
                
                if (state == AWAITING_JUMP_TABLE) {
                    if (b == 'P') {
                        uint8_t resp = 'P';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    }
                }
                else if (state == JUMP_TABLE) {
                    uint8_t cmd = b;
                    if (cmd == 2) {  // Echo
                        uint8_t echo_byte;
                        if (!read_exactly(master_fd, &echo_byte, 1)) break;
                        write(master_fd, &echo_byte, 1);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 21) {  // programmer_protocol
                        write(master_fd, "P18A", 4);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 20) {  // programmer_version
                        uint8_t resp = 3;
                        write(master_fd, &resp, 1);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 18 || cmd == 19) {  // detect chip in/out socket
                        uint8_t resp_a = 'A';
                        uint8_t resp_y = 'Y';
                        write(master_fd, &resp_a, 1);
                        write(master_fd, &resp_y, 1);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 3) {  // init programming vars
                        uint8_t vars[11];
                        if (!read_exactly(master_fd, vars, 11)) break;
                        
                        rom_size_words = (vars[0] << 8) | vars[1];
                        eeprom_size_bytes = (vars[2] << 8) | vars[3];
                        uint8_t core_type = vars[4];
                        rom_size_bytes = rom_size_words * 2;
                        
                        // Rescale buffers only if size changes
                        if (rom.size() != rom_size_bytes) {
                            rom.assign(rom_size_bytes, 0xFF);
                        }
                        if (eeprom.size() != eeprom_size_bytes) {
                            eeprom.assign(eeprom_size_bytes, 0xFF);
                        }
                        
                        // Dynamically identify chip & core family
                        if (core_type == 1 || core_type == 2 || core_type == 13) { // 16-bit PIC18
                            if (chip_id != 4640) {
                                chip_id = 4640; // PIC18F2550
                                fuses_vector.assign(7, 0xFFFF);
                                pic_id.assign(8, 0xFF);
                            }
                            std::cout << "[K150-EMULATOR-C++] Configured for 16-bit PIC18 (ChipID: " << chip_id 
                                      << ", ROM: " << rom_size_bytes << "B, EEPROM: " << eeprom_size_bytes << "B)" << std::endl;
                        } else if (core_type == 4 || core_type == 11) { // 12-bit PIC12
                            if (chip_id != 65535) {
                                chip_id = 65535; // PIC12F508
                                fuses_vector.assign(1, 0x0FFF);
                                pic_id.assign(4, 0xFF);
                            }
                            std::cout << "[K150-EMULATOR-C++] Configured for 12-bit PIC12 (ChipID: " << chip_id 
                                      << ", ROM: " << rom_size_bytes << "B)" << std::endl;
                        } else { // 14-bit PIC12/16
                            uint16_t target_id = 4192;
                            if (rom_size_words == 1024) {
                                target_id = 4032; // PIC12F675 (0x0FC0)
                            } else if (rom_size_words == 8192) {
                                target_id = 3616; // PIC16F877A (0x0E20)
                            } else {
                                target_id = 4192; // PIC16F628A (0x1060)
                            }
                            if (chip_id != target_id) {
                                chip_id = target_id;
                                fuses_vector.assign(1, 0x3FFF);
                                pic_id.assign(4, 0xFF);
                            }
                            std::cout << "[K150-EMULATOR-C++] Configured for 14-bit PIC12/16 (ChipID: " << chip_id 
                                      << ", ROM: " << rom_size_bytes << "B, EEPROM: " << eeprom_size_bytes << "B)" << std::endl;
                        }
                        
                        uint8_t resp = 'I';
                        write(master_fd, &resp, 1);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 4) {  // Turn programming voltage ON
                        uint8_t resp = 'V';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 5) {  // Turn programming voltage OFF
                        uint8_t resp = 'v';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 6) {  // Cycle programming voltages
                        uint8_t resp = 'V';
                        write(master_fd, &resp, 1);
                        state = AWAITING_JUMP_TABLE;
                    } else if (cmd == 14) {  // erase chip
                        std::fill(rom.begin(), rom.end(), 0xFF);
                        std::fill(eeprom.begin(), eeprom.end(), 0xFF);
                        uint16_t default_fuse = 0x3FFF;
                        if (chip_id == 4640) default_fuse = 0xFFFF;
                        else if (chip_id == 65535) default_fuse = 0x0FFF;
                        std::fill(fuses_vector.begin(), fuses_vector.end(), default_fuse);
                        std::fill(pic_id.begin(), pic_id.end(), 0xFF);
                        std::cout << "[K150-EMULATOR-C++] Memory erased (mock)" << std::endl;
                        uint8_t resp = 'Y';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 7) {  // program ROM
                        uint8_t wcount_bytes[2];
                        if (!read_exactly(master_fd, wcount_bytes, 2)) break;
                        uint16_t word_count = (wcount_bytes[0] << 8) | wcount_bytes[1];
                        uint8_t resp = 'Y';
                        write(master_fd, &resp, 1);
                        size_t bytes_to_read = word_count * 2;
                        std::cout << "[K150-EMULATOR-C++] Programming ROM: writing " << word_count << " words (" << bytes_to_read << " bytes)..." << std::endl;
                        for (size_t i = 0; i < bytes_to_read; i += 32) {
                            size_t chunk_size = std::min((size_t)32, bytes_to_read - i);
                            uint8_t chunk[32];
                            if (!read_exactly(master_fd, chunk, chunk_size)) break;
                            std::memcpy(&rom[i], chunk, chunk_size);
                            write(master_fd, &resp, 1);
                        }
                        uint8_t table_resp = 'P';
                        write(master_fd, &table_resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 8) {  // program EEPROM
                        uint8_t bcount_bytes[2];
                        if (!read_exactly(master_fd, bcount_bytes, 2)) break;
                        uint16_t byte_count = (bcount_bytes[0] << 8) | bcount_bytes[1];
                        uint8_t resp = 'Y';
                        write(master_fd, &resp, 1);
                        std::cout << "[K150-EMULATOR-C++] Programming EEPROM: writing " << byte_count << " bytes..." << std::endl;
                        for (size_t i = 0; i < byte_count; i += 2) {
                            uint8_t chunk[2];
                            if (!read_exactly(master_fd, chunk, 2)) break;
                            std::memcpy(&eeprom[i], chunk, 2);
                            write(master_fd, &resp, 1);
                        }
                        uint8_t extra[2];
                        if (!read_exactly(master_fd, extra, 2)) break;
                        uint8_t table_resp = 'P';
                        write(master_fd, &table_resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 9) {  // program ID and fuses
                        uint8_t body[24];
                        if (!read_exactly(master_fd, body, 24)) break;
                        if (fuses_vector.size() > 1) { // PIC18
                            std::memcpy(pic_id.data(), &body[2], 8);
                            for (size_t i = 0; i < 7; ++i) {
                                fuses_vector[i] = body[10 + i * 2] | (body[11 + i * 2] << 8);
                            }
                        } else { // PIC12/16
                            std::memcpy(pic_id.data(), &body[2], 4);
                            fuses_vector[0] = body[10] | (body[11] << 8);
                        }
                        std::cout << "[K150-EMULATOR-C++] Programmed ID: ";
                        for (size_t i = 0; i < pic_id.size(); ++i) printf("%02X", pic_id[i]);
                        printf(", Fuses: ");
                        for (size_t i = 0; i < fuses_vector.size(); ++i) printf("%04X ", fuses_vector[i]);
                        printf("\n");
                        uint8_t resp = 'Y';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    } else if (cmd == 11) {  // read ROM
                        std::cout << "[K150-EMULATOR-C++] Reading ROM (" << rom.size() << " bytes)..." << std::endl;
                        write(master_fd, rom.data(), rom.size());
                        state = JUMP_TABLE;
                    } else if (cmd == 12) {  // read EEPROM
                        std::cout << "[K150-EMULATOR-C++] Reading EEPROM (" << eeprom.size() << " bytes)..." << std::endl;
                        write(master_fd, eeprom.data(), eeprom.size());
                        state = JUMP_TABLE;
                    } else if (cmd == 13) {  // read config
                        std::cout << "[K150-EMULATOR-C++] Reading config..." << std::endl;
                        uint8_t config_header = 'C';
                        write(master_fd, &config_header, 1);
                        
                        // Build config bytes (26 bytes)
                        uint8_t config_bytes[26];
                        config_bytes[0] = chip_id & 0xFF;
                        config_bytes[1] = (chip_id >> 8) & 0xFF;
                        
                        // Copy ID (up to 8 bytes, padded with 0xFF)
                        for (size_t i = 0; i < 8; ++i) {
                            config_bytes[2 + i] = (i < pic_id.size()) ? pic_id[i] : 0xFF;
                        }
                        
                        // Copy 7 fuses (14 bytes, padded with 0xFFFF)
                        for (size_t i = 0; i < 7; ++i) {
                            uint16_t val = (i < fuses_vector.size()) ? fuses_vector[i] : 0xFFFF;
                            config_bytes[10 + i * 2] = val & 0xFF;
                            config_bytes[11 + i * 2] = (val >> 8) & 0xFF;
                        }
                        
                        config_bytes[24] = calibrate & 0xFF;
                        config_bytes[25] = (calibrate >> 8) & 0xFF;
                        
                        write(master_fd, config_bytes, 26);
                        state = JUMP_TABLE;
                    } else if (cmd == 24 || cmd == 25) {  // program cal data for 10Fxxx
                        uint8_t cal_data[4];
                        if (!read_exactly(master_fd, cal_data, 4)) break;
                        uint16_t cal = cal_data[1] | (cal_data[0] << 8);
                        uint16_t backup_cal = cal_data[3] | (cal_data[2] << 8);
                        std::cout << "[K150-EMULATOR-C++] Program 10F Calibration: 0x" << std::hex << cal 
                                  << ", Backup: 0x" << backup_cal << std::dec << std::endl;
                        uint8_t resp = 'Y';
                        write(master_fd, &resp, 1);
                        state = JUMP_TABLE;
                    } else {
                        std::cout << "[K150-EMULATOR-C++] Warning: Unhandled command " << (int)cmd << std::endl;
                        state = JUMP_TABLE;
                    }
                }
            }
        } catch (...) {
            std::cout << "[K150-EMULATOR-C++] Exception occurred. Resetting connection." << std::endl;
        }
        
        if (slave_fd >= 0) {
            close(slave_fd);
        }
        close(master_fd);
    }
}

int main(int argc, char* argv[]) {
    // Default symlink path
    std::string symlink_path = "./vtty";
    if (argc > 1) {
        symlink_path = argv[1];
    } else {
        // Resolve absolute path in directory of executable
        char result[PATH_MAX];
        ssize_t count = readlink("/proc/self/exe", result, PATH_MAX);
        if (count != -1) {
            std::string path(result, count);
            size_t pos = path.find_last_of("/");
            if (pos != std::string::npos) {
                symlink_path = path.substr(0, pos) + "/vtty";
            }
        }
    }
    
    run_emulator(symlink_path);
    return 0;
}

File: emulador_pic_k150/emulador_k150.py
#!/usr/bin/env python3
import os
import pty
import select
import struct
import sys
import time
import enum
import termios

# Response enums
class ResponseEnum(enum.Enum):
    YES = b'Y'
    NO = b'N'
    INITIALIZED = b'I'
    PROGRAMING_VOLTAGE_ENABLED = b'V'
    PROGRAMING_VOLTAGE_DISABLED = b'v'
    AT_COMMAND_JUMP_TABLE = b'P'
    INCORRECT_BYTES = b'B'
    WAITING_FOR_ACTION = b'A'
    WAITING_FOR_COMMAND = b'Q'

class HeaderEnum(enum.Enum):
    CONFIGURATION = b'C'
    PROGRAMMER_VERSION = b'B'

def read_exactly(fd, n):
    data = b''
    while len(data) < n:
        r, _, _ = select.select([fd], [], [], 2.0)
        if not r:
            raise TimeoutError("Timeout reading from virtual port")
        chunk = os.read(fd, n - len(data))
        if not chunk:
            raise ConnectionAbortedError("Connection closed")
        data += chunk
    return data

def run_emulator():
    # Setup paths
    base_dir = os.path.dirname(os.path.abspath(__file__))
    symlink_path = os.path.join(base_dir, "vtty")
    
    # Initialize PIC16F628A mock memory by default
    rom_size_words = 2048
    rom_size_bytes = rom_size_words * 2 # 4096
    eeprom_size_bytes = 128
    
    rom = bytearray(b'\xff' * rom_size_bytes)
    eeprom = bytearray(b'\xff' * eeprom_size_bytes)
    fuses = [0x3fff]
    pic_id = bytearray(b'\xff\xff\xff\xff')
    calibrate = 0x3fff
    chip_id = 4192  # PIC16F628A (0x1060)

    def get_config_bytes():
        padded_id = pic_id.ljust(8, b'\xff')
        padded_fuses = fuses + [0xffff] * (7 - len(fuses))
        return struct.pack('<H8s7HH',
            chip_id,
            padded_id,
            *padded_fuses,
            calibrate
        )

    print("================ K150 PROGRAMMER EMULATOR ================")
    print("Simulating PIC microcontrollers (Dynamic Family Support)")
    print("Press Ctrl+C to exit emulator")
    print("==========================================================")

    while True:
        # Create a new pty pair
        master, slave = pty.openpty()
        slave_name = os.ttyname(slave)
        
        # Configure slave to be RAW (no ECHO, no ICANON, no flow control)
        attrs = termios.tcgetattr(slave)
        attrs[0] = attrs[0] & ~termios.IXON & ~termios.IXOFF & ~termios.IXANY
        attrs[1] = attrs[1] & ~termios.OPOST
        attrs[3] = attrs[3] & ~termios.ECHO & ~termios.ICANON & ~termios.IEXTEN & ~termios.ISIG
        termios.tcsetattr(slave, termios.TCSANOW, attrs)
        
        # Update symlink
        if os.path.exists(symlink_path):
            os.unlink(symlink_path)
        os.symlink(slave_name, symlink_path)
        
        print(f"[K150-EMULATOR] Active virtual port: {slave_name}")
        print(f"[K150-EMULATOR] Symlink updated: {symlink_path}")
        print(f"[K150-EMULATOR] Ready. Virtual port symlink: {symlink_path}")
        print("[K150-EMULATOR] Waiting for client connection...")
        
        # Block on read from master PTY
        os.set_blocking(master, True)
        
        state = "AWAITING_JUMP_TABLE"
        buf = b''
        
        try:
            while True:
                if not buf:
                    buf = os.read(master, 1)
                    if not buf:
                        print("[K150-EMULATOR] Client disconnected (EOF).")
                        break
                
                b = buf[0]
                buf = buf[1:] # Consume 1 byte
                
                # Exit command (b'\x01') always resets state and replies 'Q'
                if b == 1:
                    os.write(master, ResponseEnum.WAITING_FOR_COMMAND.value)
                    state = "AWAITING_JUMP_TABLE"
                    continue
                
                if state == "AWAITING_JUMP_TABLE":
                    if b == ord('P'):
                        os.write(master, ResponseEnum.AT_COMMAND_JUMP_TABLE.value)
                        state = "JUMP_TABLE"
                
                elif state == "JUMP_TABLE":
                    cmd = b
                    if cmd == 2:  # Echo
                        echo_byte = read_exactly(master, 1)
                        os.write(master, echo_byte)
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 21:  # programmer_protocol
                        os.write(master, b'P18A')
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 20:  # programmer_version
                        os.write(master, b'\x03')
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 18 or cmd == 19:  # detect chip in/out socket
                        os.write(master, ResponseEnum.WAITING_FOR_ACTION.value)
                        os.write(master, ResponseEnum.YES.value)
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 3:  # init programming vars
                        vars_bytes = read_exactly(master, 11)
                        rom_size_words, eeprom_size_bytes, core_type = struct.unpack('>HHB', vars_bytes[0:5])
                        rom_size_bytes = rom_size_words * 2
                        
                        # Rescale buffers only if size changes
                        if len(rom) != rom_size_bytes:
                            rom = bytearray(b'\xff' * rom_size_bytes)
                        if len(eeprom) != eeprom_size_bytes:
                            eeprom = bytearray(b'\xff' * eeprom_size_bytes)
                        
                        # Dynamically identify chip & core family
                        if core_type in [1, 2, 13]: # 16-bit cores (PIC18)
                            if chip_id != 4640:
                                chip_id = 4640 # PIC18F2550
                                fuses = [0xffff] * 7
                                pic_id = bytearray(b'\xff' * 8)
                            print(f"[K150-EMULATOR] Configured for 16-bit PIC18 (ChipID: {chip_id}, ROM: {rom_size_bytes}B, EEPROM: {eeprom_size_bytes}B)")
                        elif core_type in [4, 11]: # 12-bit cores (PIC12)
                            if chip_id != 65535:
                                chip_id = 65535 # PIC12F508
                                fuses = [0x0fff]
                                pic_id = bytearray(b'\xff' * 4)
                            print(f"[K150-EMULATOR] Configured for 12-bit PIC12 (ChipID: {chip_id}, ROM: {rom_size_bytes}B)")
                        else: # 14-bit cores (PIC12/16)
                            target_id = 4192
                            if rom_size_words == 1024:
                                target_id = 4032 # PIC12F675 (0x0FC0)
                            elif rom_size_words == 8192:
                                target_id = 3616 # PIC16F877A (0x0E20)
                            else:
                                target_id = 4192 # PIC16F628A (0x1060)
                            if chip_id != target_id:
                                chip_id = target_id
                                fuses = [0x3fff]
                                pic_id = bytearray(b'\xff' * 4)
                            print(f"[K150-EMULATOR] Configured for 14-bit PIC12/16 (ChipID: {chip_id}, ROM: {rom_size_bytes}B, EEPROM: {eeprom_size_bytes}B)")
                            
                        # Handshake initialized response
                        os.write(master, ResponseEnum.INITIALIZED.value)
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 4:  # Turn programming voltage ON
                        os.write(master, ResponseEnum.PROGRAMING_VOLTAGE_ENABLED.value)
                        state = "JUMP_TABLE"
                    elif cmd == 5:  # Turn programming voltage OFF
                        os.write(master, ResponseEnum.PROGRAMING_VOLTAGE_DISABLED.value)
                        state = "JUMP_TABLE"
                    elif cmd == 6:  # Cycle programming voltages
                        os.write(master, ResponseEnum.PROGRAMING_VOLTAGE_ENABLED.value)
                        state = "AWAITING_JUMP_TABLE"
                    elif cmd == 14:  # erase chip
                        rom = bytearray(b'\xff' * len(rom))
                        eeprom = bytearray(b'\xff' * len(eeprom))
                        if chip_id == 4640:
                            fuses = [0xffff] * len(fuses)
                        elif chip_id == 65535:
                            fuses = [0x0fff] * len(fuses)
                        else:
                            fuses = [0x3fff] * len(fuses)
                        pic_id = bytearray(b'\xff' * len(pic_id))
                        print("[K150-EMULATOR] Memory erased (mock)")
                        os.write(master, ResponseEnum.YES.value)
                        state = "JUMP_TABLE"
                    elif cmd == 7:  # program ROM
                        word_count_bytes = read_exactly(master, 2)
                        word_count, = struct.unpack('>H', word_count_bytes)
                        os.write(master, ResponseEnum.YES.value)
                        bytes_to_read = word_count * 2
                        print(f"[K150-EMULATOR] Programming ROM: writing {word_count} words ({bytes_to_read} bytes)...")
                        for i in range(0, bytes_to_read, 32):
                            chunk = read_exactly(master, min(32, bytes_to_read - i))
                            rom[i:i+len(chunk)] = chunk
                            os.write(master, ResponseEnum.YES.value)
                        os.write(master, ResponseEnum.AT_COMMAND_JUMP_TABLE.value)
                        state = "JUMP_TABLE"
                    elif cmd == 8:  # program EEPROM
                        byte_count_bytes = read_exactly(master, 2)
                        byte_count, = struct.unpack('>H', byte_count_bytes)
                        os.write(master, ResponseEnum.YES.value)
                        print(f"[K150-EMULATOR] Programming EEPROM: writing {byte_count} bytes...")
                        for i in range(0, byte_count, 2):
                            chunk = read_exactly(master, 2)
                            eeprom[i:i+2] = chunk
                            os.write(master, ResponseEnum.YES.value)
                        read_exactly(master, 2)  # Extra two bytes
                        os.write(master, ResponseEnum.AT_COMMAND_JUMP_TABLE.value)
                        state = "JUMP_TABLE"
                    elif cmd == 9:  # program ID and fuses
                        body = read_exactly(master, 24)
                        if len(fuses) > 1: # PIC18
                            pic_id = bytearray(body[2:10]) # 8 bytes ID
                            for idx in range(7):
                                fuses[idx], = struct.unpack('<H', body[10 + idx*2 : 12 + idx*2])
                        else: # PIC12/16
                            pic_id = bytearray(body[2:6]) # 4 bytes ID
                            fuse_val, = struct.unpack('<H', body[10:12])
                            fuses[0] = fuse_val
                        print(f"[K150-EMULATOR] Programmed ID: {pic_id.hex()}, Fuses: {[f'{x:04X}' for x in fuses]}")
                        os.write(master, ResponseEnum.YES.value)
                        state = "JUMP_TABLE"
                    elif cmd == 11:  # read ROM
                        print(f"[K150-EMULATOR] Reading ROM ({len(rom)} bytes)...")
                        os.write(master, rom)
                        state = "JUMP_TABLE"
                    elif cmd == 12:  # read EEPROM
                        print(f"[K150-EMULATOR] Reading EEPROM ({len(eeprom)} bytes)...")
                        os.write(master, eeprom)
                        state = "JUMP_TABLE"
                    elif cmd == 13:  # read config
                        print("[K150-EMULATOR] Reading config...")
                        os.write(master, HeaderEnum.CONFIGURATION.value)
                        os.write(master, get_config_bytes())
                        state = "JUMP_TABLE"
                    elif cmd == 24 or cmd == 25:  # program cal data for 10Fxxx
                        cal_data = read_exactly(master, 4)
                        cal, backup_cal = struct.unpack('>HH', cal_data)
                        print(f"[K150-EMULATOR] Program 10F Calibration: 0x{cal:04X}, Backup: 0x{backup_cal:04X}")
                        os.write(master, ResponseEnum.YES.value)
                        state = "JUMP_TABLE"
                    else:
                        print(f"[K150-EMULATOR] Warning: Unhandled command {cmd}")
                        state = "JUMP_TABLE"
        except (OSError, ConnectionAbortedError) as e:
            print(f"[K150-EMULATOR] Client disconnected (Error: {e}).")
        finally:
            os.close(master)
            os.close(slave)

if __name__ == '__main__':
    try:
        run_emulator()
    except KeyboardInterrupt:
        print("\n[K150-EMULATOR] Exiting emulator. Goodbye!")
        sys.exit(0)
